### DISCUSSION TITLE
xteve streamlink hack for comparison to cabernet

### DISCUSSION THREAD

**Discussion by merrychrishna** [cabernetwork/cabernet#41] [General]
@rocky4546 please try this out. there are none of those problematic ads segemnt to be filtered. all you see is the plutotv splash screen or screensaver during ccommercials. its like those segments we were filtering dont even exist. theres no freezing or pausing and no black screen and no buffering. it just plays like it used to but without filtering.

XTEVE SETTINGS for streamlink for pluto tv
_________________________________________________________________________________________________

USE FFMPEG 4.3.2 (most reliable and is supported by all all custom channel programs)

----start_xteve.bat----
color 0E
:: C:\cygwin64\home\user\xteve2
:: start /MIN cmd /c xteve.exe -port 34409 -config C:\Users\User0\.xteve2
start /MIN cmd /c xteve.exe -port 34409

----xteve settings----
PLAYLIST
playlist tuners = 1

SETTINGS
number of tuners = 10
EPG source = XEPG

Schedule for updating (Playlist, XMLTV, Backup): = 0600,1100,1600,2100,0200
Stream Buffer:= ffmpeg or VLC                                                          <--MINE IS SET TO VLC this is important to use the streamlink hack. this will select the VLC options below that are to be replaced with streamlink parameters.
Location for the temporary files: C:\Users\User0\AppData\Local\Temp\xteve\            <--SETTING THIS TO XTEVE2 IS FOR USING XTEVE AS A SECOND INSTANCE along with changing the xteve port
Buffer Size: = 2mb
Timeout for new client connections: = 500
User Agent: = (blank)                                                                    <--THIS IS IMPORTANT OR THE STREAM WONT START
USE FFMPEG 4.3.2 (most reliable and is supported by all all custom channel programs)     <--- might not be important
FFmpeg Binary Path: = C:\Program Files (x86)\Streamlink\bin\streamlink.exe
FFmpeg Options: = --stdout --loglevel info [URL] --default-stream 570k --force --hls-segment-threads 4    <-----NOT IMPORTANT SINCE WE ARE USING VLC OPTIONS INSTEAD
VLC / CVLC Binary Path: = C:\Program Files (x86)\Streamlink\bin\streamlink.exe
VLC / CVLC Options: = --stdout --loglevel info [URL] --default-stream 1000k --force       <---THIS IS IMPORTANT its the streamlink hack parameters
Backup
Location for automatic backups:	C:\Users\User0\.xteve\backup\                         <--SETTING THIS TO .XTEVE2 IS FOR USING XTEVE AS A SECOND INSTANCE along with changing the xteve port

as a reference, the pluto tv quality choices are the following.
[cli][info] Available streams: 570k (worst), 1000k, 1500k, 2100k, 3100k (best)

references (i didnt make these):
https://github.com/evoactivity/PlutoIPTV/issues/7
https://github.com/SCP002/xTeVe
https://github.com/streamlink/streamlink/issues/3462

plutotv on xteve with streamlink (i didnt make this): 
https://github.com/ReenigneArcher/web2tv

my batch/cmd file command to run the plutotv.py script that downloads/refreshes the m3u and xmltv files in streamlink format:
python C:\cygwin64\home\user\xteve2\web2tv-main\plutotv.py -e 12 -x C:\cygwin64\home\user\xteve2\streamlinkPlutoEpg.xml --xml -m C:\cygwin64\home\user\xteve2\streamlinkPlutoM3U.m3u --m3u --streamlink

more info:
these are my settings to have my own pluto tv streams in emby through xteve through streamlink
and you need the streamlink pluto tv python file that creates the pluto tv m3u and xmltv. and that file needs to be run every hour or so at least. could be longer though. i made a .sh file which is a bash file, a linux file that runs the pluto python file i was talking about, runs it over and over on a timer.
but i have all that i can give you.
im just still testing the settings. i think its as good as i can get. i think my problem with pluto streams freezing was partly a bandwidth issue since im at my dads using 3mbps for emby.
i noticed that if you change those streamlink parameters, they wont take effect until the current movie thats plying, ends.
thats because emby is still transcoding the movie and the only way to stop transcoding is to wait for the movie to end and transcoding will just stop on its own at the end of the movie.

a simpler explaination:
Theres a python file I have called plutotv.py. You run a long python command in the command prompt to use that file. It will connect to Pluto tv servers and download the m3u file that has all the URLs for the Pluto tv channels. It also, at the same time will create a xmltv file which has all the tv guide info for each channel. When you use those they eventually will run out of guide info. So you have to keep running the long command to update your m3u and xmltv file. 
But I made a script that will run the command over and over on a timer.

And those m3u file and xmltv file are also formatted to run using streamlink as a work around to the channels freezing at commercials. Streamlink stitches the movies to the commercials so it's a smooth transition.
So in xteve, where it says VLC path and VLC options in settings, you replace the vlc path with your streamlink exe path. And change vlc options to your streamlink command that I provided
Instead of using vlc
That's like a way to hack xteve to use streamlink instead of vlc
I have the video quality for streaming set to the lowest setting which is 570k in the streamlink command because I'm at my dad's using only 3mbps in my emby client.
The video stream is kind of blurry though so you might want to try higher quality settings in the streamlink command. 

But the problem with changing the quality is the setting will not take effect until the emby transcoding has stopped for that channel because emby is still transcoding using ffmpeg and I was told by JD there's no way to stop or cancel transcoding once it's started. 
So the only way to get transcoding to stop is to let emby finish transcoding the entire movie so it will stop automatically.

You you just have to play the Pluto tv channel even if it's buffering and leave it frozen until the movie finishes.

you can change the quality to a higher quality in the streamlink command, you just have to wait till the movie is done.
Well, I mean change it before it's done, then wait till the movie is done for the quality to change on its own now that the setting has taken effect

[cli][info] Available streams: 570k (worst), 1000k, 1500k, 2100k, 3100k (best) 
These are the video qualities you can choose

**Comment by merrychrishna** [comment]
you might need to update the plutotv streamlink plugin with the new one that was just released. i suggest to either allow cabernet to use streamlink and allow to specify streamlink paraameters, or have a look at the streamlink pluto plugin to see how you can encorporate the new changes into cabernet.

**Comment by rocky4546** [comment]
Cabernet settings can be set to call a script instead of streamlink directly, allowing the user to have fun and mess with the parameters.  No changes to Cabernet are needed.  Closing discussion.
